src/interface/sparse_pack.hpp

```diff
@@ -90,6 +90,7 @@ KOKKOS_INLINE_FUNCTION PackIdx operator+(T offset, PackIdx idx) {
 // device
 namespace variable_names {
 // Struct that all variable_name types should inherit from
+constexpr int ANYDIM = -1234;
```

Comment:
I'm not too familiar with this variable name approach, so it's not 100% clear to me what `ANYDIM` is used for and why it's defined in Parthenon, when it's not being used expect the test.
Would be great if you could add a brief explanation there (also on the choice of the `-1234` magic number).

Reply:
To my understanding, the only purpose of ANYDIM is to highlight that first dim does not matter in our calculations and can be anything. We chose it to be negative number for purpose and 1234 is just random (i think). @Yurlungur might have a better explanation for this.

Reply:
I wrote a bit of an explanation above... but just like with how in C-style arrays, the size of the slowest-moving index can be free floating when you pass a multi-D array into a function, the indexing of a tensor variable doesn't depend on the size of the slowest moving index. This is a way of declaring the variable is of the correct tensor rank, but has a free-floating slowest-moving index. e.g.,
```C++
myvar : base_t<ANYDIM, 3, 3>
```
is a rank 3 tensor, which is Nx3x3, where N can be set at runtime. The reason for `-1234` is just that it's a negative number. It could be anything.

We should maybe add a comment explaining this @mari2895 . We could also add a check to the constructor that `ANYDIM` is only ever the slowest-moving index (i.e., the first one). Could probably be done with a fold expression.... Add the following function in private scope of the class (near `GetIndex`):
```C++
template<typename... Tail, REQUIRES(all_implement<integral(Tail...)>::value>
static constexpr bool CheckArgs_(int head, Tail... tail) {
    return (... && (tail > 0));
}
```
then change the constructor to:
```C++
 template <typename... Args, REQUIRES(all_implement<integral(Args...)>::value),
            REQUIRES(sizeof...(Args) == sizeof...(NCOMP))>
  KOKKOS_INLINE_FUNCTION explicit base_t(Args... args)
      : idx(GetIndex_(std::forward<Args>(args)...)) {
            static_assert(CheckArgs_(NCOMP...), 
    "All dimensions must be strictly positive, "
    "except the first (slowest), which may be ANYDIM.");
      }
```
As always, check that tests pass, run formatting, and linting.

Reply:
Done, all tests passed, formating and linting run.

Reply:
@pgrete does the revised version work for you?

Reply:
> Is the only purpose of ANYDIM just to make it a three dimensional (so that the pointer approach under the hood works, which allows to specify the first dimension at runtime)?

That's essentially right. 

> Similarly, my original question was also aimed at why ANYDIM is required at this place (i.e., part of the public interface) when it's only used in a test. 

We don't NEED `ANYDIM` at all. You can put any number you want in the slowest moving index (so long as it's positive) and indexing will work properly. However, it's a clarity thing. However, it's in the tests because I wanted to test that the machinery works with this piece in place.

> in other words, do I have to use ANYDIM on the downstream side (similar to the test) when I want to use sparse packs

No, you can use the specific tensor shape you want without ANYDIM. Or for scalars and vectors, you can ignore the indexing machinery entirely. 

> could ANYDIM also completely live in the test below (and not being part of the general interface).

It could, but I would prefer not. It's an entirely optional piece of parthenon machinery. But it's useful to support, e.g., things like multi-group moments, where the `ANYDIM` index is the number of groups.